hibernate-reactive-routes-quickstart/src/main/java/org/acme/hibernate/reactive/FruitsRoutes.java

```diff
@@ -1,123 +1,113 @@
-        }
-    }
-
     @Route(methods = GET, path = "/")
-    public Multi<Fruit> getAll(RoutingContext rc) throws Exception {
-        return session.createNamedQuery(Fruit.FIND_ALL, Fruit.class).getResults();
+    public Uni<List<Fruit>> getAll() throws Exception {
+        // In this case, it makes sense to return a Uni<List<Fruit>> because we return a reasonable amount of results
+        // Consider returning a Multi<Fruit> for result streams 
+        return session.createNamedQuery(Fruit.FIND_ALL, Fruit.class).getResultList();
     }
 
     @Route(methods = GET, path = "/:id")
-    public Uni<Fruit> getSingle(RoutingContext rc) {
-        final Integer id = Integer.valueOf(rc.request().getParam("id"));
-
-        return session.find(Fruit.class, id);
+    public Uni<Fruit> getSingle(@Param String id) {
+        return session.find(Fruit.class, Integer.valueOf(id));
     }
 
     @Route(methods = POST, path = "/")
-    public Uni<HttpServerResponse> create(RoutingContext rc) {
-        final String name = rc.getBodyAsJson().getString("name");
-        final Fruit entity = new Fruit(name);
-
-        return session.persist(entity)
-                        .onItem().produceUni(s -> session.flush())
-                        .onItem().apply(ignore -> httpResponse(rc, 201, entity));
+    public Uni<Fruit> create(@Body Fruit fruit, HttpServerResponse response) {
+        if (fruit == null || fruit.getId() != null) {
+            return Uni.createFrom().failure(new IllegalArgumentException("Fruit id invalidly set on request."));
+        }
+        return session.persist(fruit)
+                .onItem().transformToUni(session -> session.flush())
+                .onItem().transform(ignore -> {
+                    response.setStatusCode(201);
+                    return fruit;
+                });
     }
 
     @Route(methods = PUT, path = "/:id")
-    public Uni<HttpServerResponse> update(RoutingContext rc) {
-        final Integer id = Integer.valueOf(rc.request().getParam("id"));
-        final String name = rc.getBodyAsJson().getString("name");
-
-        return session.find(Fruit.class, id)
-                // if entity exists
-                .onItem().ifNotNull()
-                        .produceUni(entity -> {
-                                // Update the entity
-                                entity.setName(name);
-                                return session.flush()
-                                        .onItem().apply(ignore -> httpResponse(rc, 200, entity));
-                        })
-                // if entity doesn't exist
-                .onItem().ifNull()
-                        // Not found response code
-                        .continueWith(httpResponse(rc, 404));
+    public Uni<Fruit> update(@Body Fruit fruit, @Param String id) {
+        if (fruit == null || fruit.getName() == null) {
+            return Uni.createFrom().failure(new IllegalArgumentException("Fruit name was not set on request."));
```

Comment:
A curiosity, does it have to be a Uni in this case? Can't we just throw the exception?

Reply:
We can. But I think that it's better to avoid throwing an exception unless we have to...